vllm/worker/multi_step_model_runner.py

```diff
@@ -0,0 +1,500 @@
+            copy_stream: torch.cuda.Stream,
+            pinned_sampled_token_buffer: torch.Tensor) -> None:
+        """Pythonize the output. Blocking."""
+        if not self.pythonized:
+            self._pythonize_sampler_output_wait_on_event(
+                input_metadata, copy_stream, pinned_sampled_token_buffer)
+            self.pythonized = True
+
+    def maybe_pythonize(
+            self,
+            input_metadata: "MutableModelInputForGPUWithMultiStepMetadata",
+            copy_stream: torch.cuda.Stream,
+            pinned_sampled_token_buffer: torch.Tensor) -> None:
+        """Pythonize the output if ready, else return None. Non-blocking."""
+        if not self.pythonized:
+            self.pythonized = self._pythonize_sampler_output_if_event_ready(
+                input_metadata, copy_stream, pinned_sampled_token_buffer)
+
+    def _pythonize_sampler_output_wait_on_event(
+            self,
+            input_metadata: "MutableModelInputForGPUWithMultiStepMetadata",
+            copy_stream: torch.cuda.Stream,
+            pinned_sampled_token_buffer: torch.Tensor) -> None:
+        self.sampler_output_ready_event.synchronize()
+        with torch.cuda.stream(copy_stream):
+            _pythonize_sampler_output(input_metadata, self.sampler_output,
+                                      pinned_sampled_token_buffer,
+                                      self.sampled_token_ids)
+
+    def _pythonize_sampler_output_if_event_ready(
+            self,
+            input_metadata: "MutableModelInputForGPUWithMultiStepMetadata",
+            copy_stream: torch.cuda.Stream,
+            pinned_sampled_token_buffer: torch.Tensor) -> bool:
+        if self.sampler_output_ready_event.query():
+            with torch.cuda.stream(copy_stream):
+                _pythonize_sampler_output(input_metadata, self.sampler_output,
+                                          pinned_sampled_token_buffer,
+                                          self.sampled_token_ids)
+            return True
+        return False
+
+
+@dataclass(frozen=False)
+class MutableModelInputForGPUWithMultiStepMetadata(BroadcastableModelInput):
+    # actual frozen model input dataclass passed to _base_model_runner
+    frozen_model_input: Optional[ModelInputForGPUWithSamplingMetadata] = None
+    # list of model outputs for each step, may not be all pythonized
+    outputs: List[ModelOutput] = field(default_factory=list)
+    # used to pass sampled token ids from the last step to the current step for
+    # TP workers. Used to append to end of outputs and used by advance_step
+    last_sampled_token_ids: Optional[torch.Tensor] = None
+    current_step: int = 0
+    is_multi_step: bool = True
+    is_last_step: bool = False
+    is_first_multi_step: bool = False
+    step_cuda_events: List[torch.cuda.Event] = field(
+        default_factory=lambda: [torch.cuda.Event(blocking=True)] * 2)
+    num_seqs: int = -1
+    num_queries: int = -1
```

Comment:
These fields can be set in `__post_init__`.

Reply:
Due to the way we are serializing these dataclasses for broadcasting it does not really work if we have non-default fields as that makes them positional arguments during init and mess with de-serialization.

Reply:
I see. Thanks!